// bemcloud-storage/src/weixin.js
const _ = require('underscore');
const AVRequest = require('./request').request;

module.exports = function(AV) {

  AV.Weixin = function() {
    this._initialize.apply(this, arguments);
  };

  AV.Weixin.prototype = {
    _initialize: function(options) {
      this.appId = options.appId;
    },

    login: function(code, opts) {
      opts = opts || {};
      var request = AVRequest('weixin/login', null, null, 'POST', {
        code: code,
        inviter: opts.inviter,
        page: opts.page,
        wxAppId: this.appId
      });
      return request;
    },

    token: function(refresh) {
      var request = AVRequest('weixin/token', null, null, 'GET', {
        refresh: refresh ? '1' : null
      });
      return request;
    },

    jssdk: function(url) {
      var request = AVRequest('weixin/jssdk', null, null, 'GET', {
        url
      });
      return request;
    },

    sendTMessage: function(opts) {
      var request = AVRequest('weixin/sendTMessage', null, null, 'POST', opts);
      return request;
    }

  };

  _.extend(AV.Weixin, {
    /**
     * To create a new Op, call AV.Weixin._extend();
     * @private
     */
    _extend: AV._extend

  });

};
