// bemcloud-storage/dist/node/weixin.js
'use strict';

var _ = require('underscore');
var AVRequest = require('./request').request;

module.exports = function (AV) {

  AV.Weixin = function () {
    this._initialize.apply(this, arguments);
  };

  AV.Weixin.prototype = {
    _initialize: function _initialize(options) {
      this.appId = options.appId;
    },

    login: function login(code, opts) {
      opts = opts || {};
      var request = AVRequest('weixin/login', null, null, 'POST', {
        code: code,
        inviter: opts.inviter,
        page: opts.page,
        wxAppId: this.appId
      });
      return request;
    },

    token: function token(refresh) {
      var request = AVRequest('weixin/token', null, null, 'GET', {
        refresh: refresh ? '1' : null
      });
      return request;
    },

    jssdk: function jssdk(url) {
      var request = AVRequest('weixin/jssdk', null, null, 'GET', {
        url: url
      });
      return request;
    },

    sendTMessage: function sendTMessage(opts) {
      var request = AVRequest('weixin/sendTMessage', null, null, 'POST', opts);
      return request;
    }

  };

  _.extend(AV.Weixin, {
    /**
     * To create a new Op, call AV.Weixin._extend();
     * @private
     */
    _extend: AV._extend

  });
};
